src/wp-includes/media.php

```diff
@@ -5176,3 +5190,67 @@ function wp_get_webp_info( $filename ) {
 
 	return compact( 'width', 'height', 'type' );
 }
+
+/**
+ * Gets the default value to use for a `loading` attribute on an element.
+ *
+ * This function should only be called for a tag and context if lazy-loading is generally enabled.
+ *
+ * The function usually returns 'lazy', but uses certain heuristics to guess whether the current element is likely to
+ * appear above the fold, in which case it returns a boolean `false`, which will lead to the `loading` attribute being
+ * omitted on the element. The purpose of this refinement is to avoid lazy-loading elements that are within the initial
+ * viewport, which can have a negative performance impact.
+ *
+ * Under the hood, the function uses a global `$wp_content_media_count` variable which it increases every time it is
+ * called for an element within the main content. If the element is the very first content element, the `loading`
+ * attribute will be omitted. This default threshold of 1 content element to omit the `loading` attribute for can be
+ * customized using the {@see 'wp_omit_loading_attr_threshold'} filter.
+ *
+ * @since 5.9.0
+ *
+ * @global int $wp_content_media_count Content media count used to enumerate elements for the contexts 'the_content'
+ *                                     and 'the_post_thumbnail'.
+ *
+ * @param string $context Context for the element for which the `loading` attribute value is requested.
+ * @return string|bool The default `loading` attribute value. Either 'lazy', 'eager', or a boolean `false`, to indicate
+ *                     that the `loading` attribute should be skipped.
+ */
+function wp_get_loading_attr_default( $context ) {
+	global $wp_content_media_count;
+
+	// Only elements with 'the_content' or 'the_post_thumbnail' context have special handling.
+	if ( 'the_content' !== $context && 'the_post_thumbnail' !== $context ) {
+		return 'lazy';
+	}
+
+	// Only elements within the main query loop have special handling.
+	if ( is_admin() || ! in_the_loop() || ! is_main_query() ) {
+		return 'lazy';
+	}
+
+	/**
+	 * Filters the threshold for how many of the first content media elements to not lazy-load.
+	 *
+	 * For these first content media elements, the `loading` attribute will be omitted. By default, this is the case
+	 * for only the very first content media element.
+	 *
+	 * @since 5.9.0
+	 *
+	 * @param int $threshold Threshold for how many of the first content media elements to not lazy-load. Default 1.
+	 */
+	$eager_threshold = apply_filters( 'wp_omit_loading_attr_threshold', 1 );
+
+	// Increase the counter since this is a main query content element.
+	if ( ! isset( $wp_content_media_count ) ) {
+		$wp_content_media_count = 0;
+	}
+	$wp_content_media_count++;
```

Comment:
One complication with this approach: if a plugin happens to `apply_filters('the_content', get_post()->post_content)` before also doing `the_content()` (e.g. for analysis), then this could result in counts that are “thrown away”. 

I've run into this problem where I was attempting to inline styles for the first occurence of a block only to find on a site with a table of contents plugin that the styles don't show up because they were injected while applying `the_content` filters without rendering the output: https://github.com/westonruter/syntax-highlighting-code-block/issues/286.

See problematic code from that TOC plugin: https://github.com/openlab-at-city-tech/openlab/blob/cacf07eceb2addc1f431c6ffeae4e34b322c1899/wp-content/plugins/fixed-toc/frontend/class-frontend-control.php#L138-L142

Reply:
I suppose the worst case is that here is that more lazy-loaded images get served at the beginning of the content, which wouldn't be any worse than they are today. So probably not a big deal, but it is an edge case.

Reply:
@westonruter That's a fair point overall, but the code here has additional checks in place to make it less likely (while certainly not impossible) that that happens. For example, the TOC plugin you referenced runs that logic on `wp_enqueue_scripts`. With the code here, that wouldn't result in problems because it only applies when `in_the_loop()`, among other things (see the beginning of this function).